File: amber/go/adapters/storage/storage_contract_test.go
package amberstorage

import (
	"context"
	"encoding/json"
	"errors"
	"path/filepath"
	"testing"

	amber "github.com/0xsj/ingen/amber"
)

func TestStoreContract(t *testing.T) {
	t.Run("memory", func(t *testing.T) {
		runStoreContract(t, func(*testing.T) Store {
			return NewMemoryStore()
		})
	})
	t.Run("file", func(t *testing.T) {
		runStoreContract(t, func(t *testing.T) Store {
			store, err := NewFileStore(filepath.Join(t.TempDir(), "provenance.json"))
			if err != nil {
				t.Fatal(err)
			}
			return store
		})
	})
	t.Run("key-value", func(t *testing.T) {
		runStoreContract(t, func(t *testing.T) Store {
			store, err := NewKeyValueStore(NewMapKeyValueBackend(), "contract/provenance")
			if err != nil {
				t.Fatal(err)
			}
			return store
		})
	})
}

// runStoreContract is the shared semantic suite for every Store
// implementation. Values are deliberately inserted child-first so an
// implementation cannot accidentally pass by returning insertion order.
func runStoreContract(t *testing.T, newStore func(*testing.T) Store) {
	t.Helper()
	store := newStore(t)
	ctx := context.Background()

	root, err := amber.Start()
	if err != nil {
		t.Fatal(err)
	}
	retry, err := root.Retry()
	if err != nil {
		t.Fatal(err)
	}
	child, err := root.Child(amber.ChildOptions{Origin: amber.OriginIncoming})
	if err != nil {
		t.Fatal(err)
	}

	if err := store.Put(ctx, amber.Provenance{}); !errors.Is(err, amber.ErrInvalidProvenance) {
		t.Fatalf("invalid provenance error = %v", err)
	}
	for _, value := range []amber.Provenance{child, retry, root} {
		if err := store.Put(ctx, value); err != nil {
			t.Fatalf("put %s: %v", value.ExecutionID(), err)
		}
	}
	if err := store.Put(ctx, root); err != nil {
		t.Fatalf("same value should be idempotent: %v", err)
	}

	stored, err := store.Get(ctx, root.ExecutionID())
	if err != nil {
		t.Fatalf("get root: %v", err)
	}
	if stored.ExecutionID() != root.ExecutionID() {
		t.Fatalf("stored execution ID = %s, want %s", stored.ExecutionID(), root.ExecutionID())
	}

	conflicting, err := contractConflictValue(root)
	if err != nil {
		t.Fatal(err)
	}
	if err := store.Put(ctx, conflicting); !errors.Is(err, ErrConflict) {
		t.Fatalf("conflicting value error = %v", err)
	}
	stored, err = store.Get(ctx, root.ExecutionID())
	if err != nil || stored.Origin() != root.Origin() {
		t.Fatalf("conflict overwrote root: origin=%s err=%v", stored.Origin(), err)
	}

	workHistory, err := store.ListByWorkID(ctx, root.WorkID())
	if err != nil {
		t.Fatalf("work history: %v", err)
	}
	assertStoreHistory(t, "work", workHistory, root.ExecutionID(), retry.ExecutionID())
	causalHistory, err := store.ListByCausation(ctx, "execution", root.ExecutionID())
	if err != nil {
		t.Fatalf("causation history: %v", err)
	}
	assertStoreHistory(t, "causation", causalHistory, retry.ExecutionID(), child.ExecutionID())
	correlatedHistory, err := store.ListByCorrelationID(ctx, root.CorrelationID())
	if err != nil {
		t.Fatalf("correlation history: %v", err)
	}
	assertStoreHistory(t, "correlation", correlatedHistory, root.ExecutionID(), retry.ExecutionID(), child.ExecutionID())

	missingID := amber.ID("99999999-9999-4999-8999-999999999999")
	if _, err := store.Get(ctx, missingID); !errors.Is(err, ErrNotFound) {
		t.Fatalf("missing value error = %v", err)
	}

	cancelled, cancel := context.WithCancel(ctx)
	cancel()
	if err := store.Put(cancelled, root); !errors.Is(err, context.Canceled) {
		t.Fatalf("cancelled put error = %v", err)
	}
	if _, err := store.Get(cancelled, root.ExecutionID()); !errors.Is(err, context.Canceled) {
		t.Fatalf("cancelled get error = %v", err)
	}
	if _, err := store.ListByWorkID(cancelled, root.WorkID()); !errors.Is(err, context.Canceled) {
		t.Fatalf("cancelled work query error = %v", err)
	}
}

func assertStoreHistory(t *testing.T, name string, history []amber.Provenance, expected ...amber.ID) {
	t.Helper()
	if len(history) != len(expected) {
		t.Fatalf("%s history length = %d, want %d", name, len(history), len(expected))
	}
	for index, expectedID := range expected {
		if history[index].ExecutionID() != expectedID {
			t.Fatalf("%s history[%d] = %s, want %s", name, index, history[index].ExecutionID(), expectedID)
		}
	}
}

func contractConflictValue(provenance amber.Provenance) (amber.Provenance, error) {
	data, err := provenance.MarshalJSON()
	if err != nil {
		return amber.Provenance{}, err
	}
	var object map[string]any
	if err := json.Unmarshal(data, &object); err != nil {
		return amber.Provenance{}, err
	}
	object["origin"] = string(amber.OriginIncoming)
	conflictingData, err := json.Marshal(object)
	if err != nil {
		return amber.Provenance{}, err
	}
	return amber.FromJSON(conflictingData)
}

File: amber/typescript/src/storage.contract.test.ts
import {
  KeyValueStore,
  MapKeyValueBackend,
  MemoryStore,
  StorageConflictError,
} from "./storage.js";
import type { ProvenanceStore } from "./storage.js";
import { Provenance } from "./provenance.js";

function assert(condition: unknown, message: string): asserts condition {
  if (!condition) {
    throw new Error(message);
  }
}

async function expectRejected(
  action: () => Promise<void>,
  predicate: (error: unknown) => boolean,
  message: string,
): Promise<void> {
  try {
    await action();
  } catch (error) {
    if (!predicate(error)) {
      throw new Error(`${message}: unexpected error ${String(error)}`);
    }
    return;
  }
  throw new Error(`${message}: operation was accepted`);
}

async function runStoreContract(name: string, store: ProvenanceStore): Promise<void> {
  const root = Provenance.start();
  const retry = root.retry();
  const child = root.child({ origin: "incoming" });

  await expectRejected(
    () => store.put({} as Provenance),
    (error) => error instanceof TypeError,
    `${name} must reject non-Provenance values`,
  );

  // Insert out of semantic order so implementations cannot accidentally pass
  // by returning insertion order from history queries.
  for (const value of [child, retry, root]) {
    await store.put(value);
  }
  await store.put(root);

  const stored = await store.get(root.execution_id);
  assert(stored?.execution_id === root.execution_id, `${name} could not read root`);

  const conflicting = Provenance.fromJSON({ ...root.toJSON(), origin: "incoming" });
  await expectRejected(
    () => store.put(conflicting),
    (error) => error instanceof StorageConflictError,
    `${name} must reject conflicting immutable values`,
  );
  const preserved = await store.get(root.execution_id);
  assert(preserved?.origin === root.origin, `${name} allowed a conflict to overwrite root`);

  const workHistory = await store.listByWorkId(root.work_id);
  assertHistory(name, "work", workHistory, root.execution_id, retry.execution_id);

  const causalHistory = await store.listByCausation("execution", root.execution_id);
  assertHistory(name, "causation", causalHistory, retry.execution_id, child.execution_id);

  const correlationHistory = await store.listByCorrelationId(root.correlation_id);
  assertHistory(
    name,
    "correlation",
    correlationHistory,
    root.execution_id,
    retry.execution_id,
    child.execution_id,
  );

  const missingId = "99999999-9999-4999-8999-999999999999";
  const missing = await store.get(missingId);
  assert(missing === undefined, `${name} must distinguish a missing lookup`);
  assert(
    (await store.listByWorkId(missingId)).length === 0,
    `${name} must return an empty history for an unknown work ID`,
  );
}

async function runConcurrentStoreCheck(name: string, store: ProvenanceStore): Promise<void> {
  const root = Provenance.start();
  const values = Array.from({ length: 24 }, () => root.child());
  await Promise.all(values.flatMap((value) => [store.put(value), store.put(value)]));

  await Promise.all(
    Array.from({ length: 8 }, async () => {
      for (let index = 0; index < 32; index += 1) {
        await store.listByCorrelationId(root.correlation_id);
      }
    }),
  );

  const history = await store.listByCorrelationId(root.correlation_id);
  assert(
    history.length === values.length,
    `${name} concurrent writes stored ${history.length} records, want ${values.length}`,
  );
}

function assertHistory(
  name: string,
  query: string,
  history: readonly Provenance[],
  ...expectedIds: string[]
): void {
  assert(
    history.length === expectedIds.length,
    `${name} ${query} history length=${history.length}, want ${expectedIds.length}`,
  );
  for (const [index, expectedId] of expectedIds.entries()) {
    assert(
      history[index]?.execution_id === expectedId,
      `${name} ${query} history[${index}] is ${history[index]?.execution_id}, want ${expectedId}`,
    );
  }
}

await runStoreContract("memory", new MemoryStore());
await runStoreContract("key-value", new KeyValueStore(new MapKeyValueBackend(), "contract/provenance"));
await runConcurrentStoreCheck("memory", new MemoryStore());
await runConcurrentStoreCheck("key-value", new KeyValueStore(new MapKeyValueBackend(), "concurrent/provenance"));

console.log("TypeScript storage contract tests passed");
